app/controllers/Utility.scala

```diff
@@ -2,26 +2,44 @@ package controllers
 
 import com.gu.googleauth.AuthAction
 import com.gu.janus.model.JanusData
+import conf.Config
 import logic.Owners
-import play.api.Mode
+import play.api.{Configuration, Logging, Mode}
 import play.api.mvc._
 
+import scala.util.{Failure, Try}
+
 class Utility(
     janusData: JanusData,
     controllerComponents: ControllerComponents,
-    authAction: AuthAction[AnyContent]
+    authAction: AuthAction[AnyContent],
+    configuration: Configuration
 )(implicit mode: Mode, assetsFinder: AssetsFinder)
-    extends AbstractController(controllerComponents) {
+    extends AbstractController(controllerComponents)
+    with Logging {
 
   def healthcheck = Action {
     Ok("ok")
   }
 
   def accounts = authAction { implicit request =>
     val sortedAccounts = janusData.accounts.toList.sortBy(_.name.toLowerCase)
-    val owners = sortedAccounts.map(account =>
-      account -> Owners.accountOwners(account, janusData.access)
-    )
-    Ok(views.html.accounts(owners, request.user, janusData))
+    val accountData =
+      sortedAccounts.map { awsAccount =>
+        (
+          awsAccount,
+          Owners.accountPermissions(awsAccount, janusData.access),
+          Config.accountNumber(awsAccount.authConfigKey, configuration)
+        )
+      }
+    // log any account number errors we accumulated
+    accountData
+      .collect { case (account, _, Failure(err)) =>
+        (account, err)
+      }
+      .foreach { case (account, err) =>
+        logger.warn(s"Couldn't lookup account number for ${account.name}", err)
+      }
```

Comment:
Perhaps this could be broken into subfunctions for clarity but comment makes it clear what's going on so again not important

Reply:
I'm also always a bit nervous about having big controllers, particularly where it makes it hard to test the logic. This app doesn't use what i would now call the normal "for comp at the edge" style, but I agree with you this is still too far from the spirit of that!

I'll pull the logic pieces out into a testable locations so this Action is just a list of well-tested steps.